### PR TITLE
Enable dev preview in non interactive terminals

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/develop-app-management/issues/1519

### WHAT is this pull request doing?
- When you run the command in a non tty termina and the app supports `dev preview`
  - `dev preview` mode is enabled before the process are run
  - `dev preview` mode is disabled when an abort signal is detected

### How to test your changes?
- Run the `dev` with and app that includes a `theme app extension` or a `draftable` one
- Stop the interactive `dev` command and check in the partners dashboard that the `dev preview` is `off`
- Run the `dev` command as non interactive
```sh
true | (p shopify app dev --path /path/to/your/app) 2>&1 | cat
```
- Check in the partners dashboard that the `dev preview` is `on`

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: .changeset/healthy-elephants-roll.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/app': major
```
Comment (gonzaloriestra) [inline .changeset/healthy-elephants-roll.md:2]:
```suggestion
'@shopify/app': minor
```

File: packages/app/src/cli/services/dev/ui.test.tsx [hunk 2]
```diff
@@ -154,6 +157,58 @@ describe('ui', () => {
...
+      const abortController = new AbortController()
+
+      await renderDev({processes, previewUrl, graphiqlUrl, app, abortController})
+      abortController.abort()
+
+      expect(enableDeveloperPreview).toHaveBeenCalledOnce()
+      expect(disableDeveloperPreview).toHaveBeenCalledOnce()
+    })
+
+    test("not enable dev preview when terminal doesn't support TTY and the app does not supports it", async () => {
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/ui.test.tsx:186]:
```suggestion
    test("don't enable dev preview when terminal doesn't support TTY and the app doesn't support it", async () => {
```

File: packages/app/src/cli/services/dev/ui.test.tsx [hunk 3]
```diff
@@ -142,6 +144,7 @@ describe('ui', () => {
       }
 
       const abortController = new AbortController()
+      abortController.abort()
```
Comment (amcaplan) [inline packages/app/src/cli/services/dev/ui.test.tsx:147]: Why is this here in this test, but follows `renderDev` elsewhere?

File: packages/app/src/cli/services/dev/ui.tsx [hunk 4]
```diff
@@ -82,3 +88,21 @@ async function partnersURL(organizationId: string, appId: string) {
...
+async function renderDevNonInteractive({
+  processes,
+  app: {apiKey, token, canEnablePreviewMode},
+  abortController,
+}: Omit<DevProps, 'previewUrl'>) {
+  if (canEnablePreviewMode) {
+    await enableDeveloperPreview({apiKey, token})
+    abortController?.signal.addEventListener('abort', async () => {
+      await disableDeveloperPreview({apiKey, token})
+    })
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/ui.tsx:101]: Useless for Spin?

```suggestion
```

File: packages/app/src/cli/services/dev/ui.test.tsx [hunk 5]
```diff
@@ -154,6 +157,58 @@ describe('ui', () => {
...
+        token: '123',
+      }
+
+      const abortController = new AbortController()
+
+      await renderDev({processes, previewUrl, graphiqlUrl, app, abortController})
+      abortController.abort()
+
+      expect(enableDeveloperPreview).not.toHaveBeenCalledOnce()
+      expect(disableDeveloperPreview).not.toHaveBeenCalledOnce()
```
Comment (amcaplan) [inline packages/app/src/cli/services/dev/ui.test.tsx:209]: I know it's nitpicky, but wouldn't this be better as:

```suggestion
      expect(enableDeveloperPreview).not.toHaveBeenCalled()
      expect(disableDeveloperPreview).not.toHaveBeenCalled()
```

I could be wrong, but maybe `not.toHaveBeenCalledOnce()` would be truthy if it were called twice. Removing the `Once` prevents this ambiguity.

**Review by gonzaloriestra — APPROVED** [review]
Working like a charm! 👏
